### PR TITLE
feat: Noports desktop cram auth

### PR DESCRIPTION
**Pull request by xavierchanth**
**- What I did**

- Added a cram auth screen to the app
- Added a button to use cram auth instead from the otp screen
- Automatic fallback to cram screen when registrar url / api key is missing
- Added a starter flatpak setup for testing, since I'm testing from linux
- Added command-line parser to NoPorts Desktop, which takes 3 flags/options:
  - help
  - register-root - register a new root domain / registrar / apiKey into the app
  - no-include-default-roots - removes the default roots from the app
    (root.atsign.org & vip.ve.atsign.zone)
- fixed an onboarding bug in the dedicated branch of at_onboarding_flutter for NoPorts:
https://github.com/atsign-foundation/at_widgets/commit/1e244eb5f522d31d9d0eb3c67990d7d6bd612446
  - We also pass in the appropriate variables related to the bug in this function.
- Disable the `Try test profile` widget when not onboarded with a prod atSign.

When registering a root it is useful to note that only the domain and description are required. When registrarUrl / apiKey is missing for a root, the app will fallback to the new cram screen.

**- How I did it**

**- How to verify it**

Tested cram activation with prod. Tested register-root via staging.

How to test this branch:

1. Checkout the branch
2. `cd packages/dart/npt_flutter && flutter pub get`
3. Build in release mode: `flutter build <OS> --release`
4. Run the release via command-line with these arguments for NoPorts to add staging to the list of root-domains in the app:
`npt_flutter --register-root 'domain:root.atsign.wtf;description:Staging'`
You may also replace the root's values with a private environment or add another, comma separated:
`npt_flutter --register-root 'domain:my.private.root.wtf;description:My private root'`
or
```sh
npt_flutter --register-root \
'domain:root.atsign.wtf;description:Staging,domain:my.private.root.wtf;description:My private root'
```
5. Test production without OTP:
   5.1. Enter an unactivated prod atSign with prod root-domain set
   5.2. Click the option below the pin code form to "use a license key instead"
   5.3. Copy the cram key from the registrar
   5.4. Paste the cram key
   5.5. Ensure activation was successful
6. Test staging with CRAM
   6.1. Enter an unactivated staging atSign with staging root-domain set
   6.2. Copy the cram key from the registrar
   6.3. Paste the cram key
   6.4. Ensure activation was successful

Note that as of this update, there is no code in NoPorts Desktop that implies staging should work, thus staging serves as a suitable test for a private environment. There are *some* uses of the staging url in other dependencies, but they *should* be ignored at this point, since NoPorts should be passing explicit values in all of these places.

**- Description for the changelog**
feat: Noports desktop cram auth

### CODE REVIEW

File: packages/dart/npt_flutter/lib/features/onboarding/widgets/activate_atsign_dialog.dart
Comment (xavierchanth) [inline packages/dart/npt_flutter/lib/features/onboarding/widgets/activate_atsign_dialog.dart:1]: renamed to activate_otp_dialog.dart, with some minor modifications to reflect the changes in 'constants.dart'

File: packages/dart/npt_flutter/lib/features/onboarding/widgets/onboarding_button.dart [hunk 1]
```diff
@@ -204,51 +220,59 @@ class _OnboardingButtonState extends State<OnboardingButton> {
...
+                  onboardingUtil: util,
+                )
+              : ActivateOtpDialog(
+                  atSign: atsign,
+                  apiKey: apiKey,
+                  config: util.config,
+                  registrarUrl: root!.registrarUrl!,
+                  onboardingUtil: util,
+                  waitForTeapot: initialStatus != AtSignStatus.teapot,
+                ),
```
Comment (xavierchanth) [inline packages/dart/npt_flutter/lib/features/onboarding/widgets/onboarding_button.dart]: Here's where we pick the activation screen based on whether apikey & reg url are known or not.

File: packages/dart/npt_flutter/lib/localization/app_en.arb
Comment (xavierchanth) [inline packages/dart/npt_flutter/lib/localization/app_en.arb:1]: For localization:

I sorted with the unix sort command, which has changed the order in instances where: "yaml" and "yamlRecommended" occur. However it was previously sorted, had a different sort order for `"` vs alpha.

The three new entries are:
- "orSpace"
- "activateUsingLicense"
- "typePasteLicense"

File: packages/dart/npt_flutter/linux/CMakeLists.txt [hunk 2]
```diff
@@ -1,5 +1,5 @@
 # Project-level configuration.
```
Comment (xavierchanth) [inline packages/dart/npt_flutter/linux/CMakeLists.txt:1]: regenerated the linux directory with flutter create.

File: packages/dart/npt_flutter/linux/CMakeLists.txt [hunk 3]
```diff
@@ -43,7 +44,11 @@ function(APPLY_STANDARD_SETTINGS TARGET)
   target_compile_features(${TARGET} PUBLIC cxx_std_14)
   target_compile_options(${TARGET} PRIVATE -Wall -Werror)
   target_compile_options(${TARGET} PRIVATE "$<$<NOT:$<CONFIG:Debug>>:-O3>")
-  target_compile_definitions(${TARGET} PRIVATE "$<$<NOT:$<CONFIG:Debug>>:NDEBUG>")
+  target_compile_definitions(
+    ${TARGET}
+    PRIVATE "$<$<NOT:$<CONFIG:Debug>>:NDEBUG>"
+  )
+  target_compile_options(${TARGET} PRIVATE -Wno-error=deprecated-declarations) # Allow deprecated warnings for tray_manager package
```
Comment (xavierchanth) [inline packages/dart/npt_flutter/linux/CMakeLists.txt]: This line is added to allow builds despite usage of a deprecated libadwaita api in the tray_manager plugin.

File: packages/dart/npt_flutter/package-linux.sh [hunk 4]
```diff
@@ -0,0 +1,20 @@
+#!/usr/bin/env bash
```
Comment (xavierchanth) [inline packages/dart/npt_flutter/package-linux.sh:1]: This script isn't 100% finished, but my changeset for this was mixed in with the cram auth work, so I'm pushing with intention to complete at a later point.

File: packages/dart/npt_flutter/flatpak-flutter.yaml [hunk 5]
```diff
@@ -0,0 +1,42 @@
+id: com.atsign.NoPortsDesktop
```
Comment (xavierchanth) [inline packages/dart/npt_flutter/flatpak-flutter.yaml:1]: spec to generate a flatpak manifest from.

**Comment by xavierchanth** [comment]
Converted back to draft to test a few more edge cases.

**Comment by xavierchanth** [comment]
Related commit in at_onboarding_flutter: https://github.com/atsign-foundation/at_widgets/commit/1e244eb5f522d31d9d0eb3c67990d7d6bd612446

File: .vscode/launch.json [hunk 6]
```diff
@@ -48,11 +47,33 @@
...
+        "--register-root",
+        "-a",
+        "domain:root.atsign.wtf;description:Staging",
+        "-a",
+        "--register-root",
+        "-a",
+        "domain:fake.atsign.wtf;description:fake",
+        "-a",
+        "--no-include-default-roots"
+      ]
```
Comment (xavierchanth) [inline .vscode/launch.json:76]: Debugging profile against staging for NoPorts Desktop

File: packages/dart/npt_flutter/lib/features/onboarding/widgets/onboarding_button.dart [hunk 7]
```diff
@@ -204,51 +220,58 @@ class _OnboardingButtonState extends State<OnboardingButton> {
...
+                  config: util.config,
+                )
+              : ActivateOtpDialog(
+                  atSign: atsign,
+                  apiKey: apiKey,
+                  config: util.config,
+                  registrarUrl: root!.registrarUrl!,
+                  onboardingUtil: util,
+                  waitForTeapot: initialStatus != AtSignStatus.teapot,
+                ),
```
Comment (xavierchanth) [inline packages/dart/npt_flutter/lib/features/onboarding/widgets/onboarding_button.dart]: Where we determine whether to show OTP or CRAM screen immediately

File: packages/dart/npt_flutter/lib/features/onboarding/widgets/activate_otp_dialog.dart [hunk 8]
```diff
@@ -0,0 +1,305 @@
...
+  final strings = AppLocalizations.of(App.navState.currentContext!)!;
+
+  @override
+  Widget build(BuildContext context) {
+    if (_showCram) {
+      return ActivateCramDialog(
+        atSign: widget.atSign,
+        config: widget.config,
+      );
+    }
```
Comment (xavierchanth) [inline packages/dart/npt_flutter/lib/features/onboarding/widgets/activate_otp_dialog.dart]: OTP screen has an internal switch to show the cram screen instead

File: packages/dart/npt_flutter/pubspec.lock [hunk 9]
```diff
@@ -196,8 +196,8 @@ packages:
     dependency: "direct main"
     description:
       path: "packages/at_onboarding_flutter"
-      ref: at_onboarding_flutter_layers
-      resolved-ref: "3b22ed2d8608b11203d7a3fa767d5a64406eef01"
+      ref: "1e244eb5f522d31d9d0eb3c67990d7d6bd612446"
+      resolved-ref: "1e244eb5f522d31d9d0eb3c67990d7d6bd612446"
```
Comment (xavierchanth) [inline packages/dart/npt_flutter/pubspec.lock]: New commit in at_onboarding_flutter layers, I've switched to using the commit ref, because pub doesn't check the remote for new changes on a branch, nor should it to be fair. Pinning to a commit is safer, and ensures that pub fetches new changes when we tell it to.

File: packages/dart/sshnp_flutter/lib/l10n/app_localizations.dart
Comment (xavierchanth) [inline packages/dart/sshnp_flutter/lib/l10n/app_localizations.dart:1]: Whoops, please ignore this, deprecated app, so it doesn't really matter. I accidentally ran pub get in this package.

File: packages/dart/npt_flutter/lib/features/onboarding/util/activate_util.dart [hunk 10]
```diff
@@ -9,6 +9,8 @@ import 'package:at_onboarding_flutter/src/utils/at_onboarding_response_status.da
 import 'package:at_server_status/at_server_status.dart';
 import 'package:http/http.dart';
 import 'package:http/io_client.dart';
+// ignore: implementation_imports
+import 'package:at_onboarding_flutter/src/utils/at_onboarding_app_constants.dart';
```
Comment (xavierchanth) [inline packages/dart/npt_flutter/lib/features/onboarding/util/activate_util.dart:13]: This is exposed so we can set the root domain in one of the places in at_onboarding_flutter

**Review by gkc — DISMISSED** [review]
